bool dynamic_flag
bool obstacle_flag
bool flag